mozjs/js/rust/magic_codegen/src/lib.rs

```diff
@@ -0,0 +1,144 @@
+#![feature(proc_macro)]
+extern crate proc_macro;
+extern crate syn;
+#[macro_use]
+extern crate quote;
+
+use syn::{ Body, VariantData };
+use proc_macro::TokenStream;
+
+#[proc_macro_derive(MagicDom)]
+pub fn magic_dom(input: TokenStream) -> TokenStream {
+    let s = input.to_string();
+    let ast = syn::parse_macro_input(&s).unwrap();
+
+	let gen = match ast.body {
+		Body::Enum(_) => panic!("#[derive(magic-dom)] can only be used with structs"),
+		Body::Struct(ref data) => match_struct(&ast, data),
+	};
+    gen.to_string().parse().unwrap() 
+}
+
+fn match_struct( ast: &syn::DeriveInput, variant: &syn::VariantData ) -> quote::Tokens {
+    let name_spec = &ast.ident;
+    let name_spec_str = name_spec.to_string();
+    let split_name: Vec<&str> = name_spec_str.split('_').collect();
```

Comment:
Rather than allocating heap storage for the split elements and then only using the first element, we can avoid the heap allocation like so:

```rust
let name = name_spec_str.split('_').next()
    .expect("Should have a '_' in the magic dom struct name");
let name = quote::Ident::from(name);
```

This also shows off variable shadowing which is a pretty common pattern in Rust. Each `let` is its own variable binding, not the same one, so they can have different types, different mutability, etc. It is useful when converting one thing between multiple different types one step at a time.